### ISSUE TITLE
TestManifestPartitioning is extremely slow with RequestsHttpConnection

### ISSUE THREAD

**Issue by hannes-ucsc** [DataBiosphere/azul#5580]
With the patch below, `TestManifestPartitioning.test` is extremely slow ([>26min](https://github.com/DataBiosphere/azul/actions/runs/6360341171/job/17274056546#step:4:3358) vs [1min](https://github.com/DataBiosphere/azul/actions/runs/6345854461/job/17238511355#step:4:2892) on GitHub, with a similar change). The central aspect of the patch is that is configures the ES client we use with a subclass of `elasticsearch.RequestsHttpConnection` as the connection class instead of a subclass `elasticsearch.Urllib3HttpConnection`. The respective subclasses improve logging and were added in #4233. Both subclasses are inheriting the logging code from a common mixin, so the logging aspects are identical between them.

This patch makes the tests use the same ES client connection class as the application, something that is desirable because it increases makes the tests more representative of application behavior. The reason the application uses `RequestsHttpConnection` is that we have to use AWS request signing for requests against an AWS Opensearch domain, and the only available signing library supports Requests only, not the underlying Urllib3. The reason the tests use `Urllib3HttpConnection` is because that connection class is the default.

Ideally, since we're ditching Requests long-term, I would want to use `Urllib3HttpConnection` for both tests and application, but we would need to implement AWS request signing for Urllib3 or find a library that does that.

**For now, the worrying observation here is that a test is slow with the connection class that we use in the application. Does this mean that the application is slowed down unnecessarily? Does the test not work correctly?**

Note that most of the patch actually deals with complications resulting from using Requests to talk to ES. Some tests use Responses to mock Requests, and the mocks need to be told to pass-thru requests to ES, in addition to requests to the application running locally.

```diff
Index: test/health_check_test_case.py
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/test/health_check_test_case.py b/test/health_check_test_case.py
--- a/test/health_check_test_case.py	(revision 848a45385e8689310693a8e1668074d0fd463770)
+++ b/test/health_check_test_case.py	(date 1696057914565)
@@ -286,8 +286,7 @@
             return requests.get(str(self.base_url.set(path=path)))
 
     def helper(self):
-        helper = responses.RequestsMock()
-        helper.add_passthru(str(self.base_url))
+        helper = self._requests_mock_helper()
         # We originally shared the Requests mock with Moto which had this set
         # to False. Because of that, and without noticing, we ended up mocking
         # more responses than necessary for some of the tests. Instead of
Index: test/service/test_drs.py
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/test/service/test_drs.py b/test/service/test_drs.py
--- a/test/service/test_drs.py	(revision 848a45385e8689310693a8e1668074d0fd463770)
+++ b/test/service/test_drs.py	(date 1696057914566)
@@ -69,8 +69,7 @@
         return ChaliceConfig.create(lambda_timeout=15)
 
     def _get_data_object(self, file_uuid, file_version):
-        with responses.RequestsMock() as helper:
-            helper.add_passthru(str(self.base_url))
+        with self._requests_mock_helper() as helper:
             drs_url = dss_dos_object_url(file_uuid=file_uuid,
                                          catalog=self.catalog,
                                          file_version=file_version,
@@ -157,8 +156,7 @@
         file_version = '2018-11-02T113344.698028Z'
         for redirects in (0, 1, 2, 6):
             with self.subTest(redirects=redirects):
-                with responses.RequestsMock() as helper:
-                    helper.add_passthru(str(self.base_url))
+                with self._requests_mock_helper() as helper:
                     self._mock_responses(helper, redirects, file_uuid, file_version=file_version)
                     # Make first client request
                     url = dss_drs_object_url(file_uuid=file_uuid,
@@ -280,7 +278,6 @@
 
     def _mock_responses(self, helper, redirects, file_uuid, file_version=None):
         assert redirects >= 0
-        helper.add_passthru(str(self.base_url))
         if redirects == 0:
             helper.add(self._dss_response(file_uuid, file_version, 'aws', initial=True, _301=False))
             helper.add(self._dss_response(file_uuid, file_version, 'gcp', initial=True, _301=False))
@@ -299,8 +296,7 @@
     def test_data_object_not_found(self):
         file_uuid = 'NOT_A_GOOD_IDEA'
         error_body = 'DRS should just proxy the DSS for error responses'
-        with responses.RequestsMock() as helper:
-            helper.add_passthru(str(self.base_url))
+        with self._requests_mock_helper() as helper:
             url = f'{config.dss_endpoint}/files/{file_uuid}'
             helper.add(responses.Response(method=responses.GET,
                                           body=error_body,
Index: test/app_test_case.py
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/test/app_test_case.py b/test/app_test_case.py
--- a/test/app_test_case.py	(revision 848a45385e8689310693a8e1668074d0fd463770)
+++ b/test/app_test_case.py	(date 1696057914564)
@@ -88,6 +88,9 @@
         host, port = self.server_thread.address
         return furl(scheme='http', host=host, port=port)
 
+    def _requests_mock_passthru_urls(self):
+        return super()._requests_mock_passthru_urls() + [self.base_url]
+
     @classmethod
     def setUpClass(cls):
         super().setUpClass()
Index: test/azul_test_case.py
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/test/azul_test_case.py b/test/azul_test_case.py
--- a/test/azul_test_case.py	(revision 848a45385e8689310693a8e1668074d0fd463770)
+++ b/test/azul_test_case.py	(date 1696057914565)
@@ -30,8 +30,12 @@
     Credentials,
 )
 import botocore.session
+from furl import (
+    furl,
+)
 import moto.backends
 import moto.core.models
+import responses
 
 from azul import (
     CatalogName,
@@ -372,6 +376,15 @@
         cls._drs_domain_mock.stop()
         cls._drs_domain_mock = None
 
+    def _requests_mock_helper(self) -> responses.RequestsMock:
+        helper = responses.RequestsMock()
+        for url in self._requests_mock_passthru_urls():
+            helper.add_passthru(str(url))
+        return helper
+
+    def _requests_mock_passthru_urls(self) -> list[furl]:
+        return []
+
 
 class CatalogTestCase(AzulUnitTestCase, metaclass=ABCMeta):
     catalog: CatalogName = 'test'
Index: test/service/test_repository_proxy.py
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/test/service/test_repository_proxy.py b/test/service/test_repository_proxy.py
--- a/test/service/test_repository_proxy.py	(revision 848a45385e8689310693a8e1668074d0fd463770)
+++ b/test/service/test_repository_proxy.py	(date 1696057914566)
@@ -299,8 +299,7 @@
                                                  ('foo bar.txt', 'grbM6udwp0n/QE/L/RYfjtQCS/U='),
                                                  ('foo&bar.txt', 'r4C8YxpJ4nXTZh+agBsfhZ2e7fI=')]:
                         with self.subTest(fetch=fetch, file_name=file_name, wait=wait):
-                            with responses.RequestsMock() as helper:
-                                helper.add_passthru(str(self.base_url))
+                            with self._requests_mock_helper() as helper:
                                 fixed_time = 1547691253.07010
                                 expires = str(round(fixed_time + 3600))
                                 s3_url = furl(url=f'https://{bucket_name}.s3.amazonaws.com',
Index: test/es_test_case.py
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/test/es_test_case.py b/test/es_test_case.py
--- a/test/es_test_case.py	(revision 848a45385e8689310693a8e1668074d0fd463770)
+++ b/test/es_test_case.py	(date 1696057914565)
@@ -4,6 +4,10 @@
     mock,
 )
 
+from furl import (
+    furl,
+)
+
 from azul import (
     config,
 )
@@ -53,6 +57,11 @@
             cls._kill_containers()
             raise
 
+    def _requests_mock_passthru_urls(self):
+        host, port = config.es_endpoint
+        url = furl(scheme='http', host=host, port=port)
+        return super()._requests_mock_passthru_urls() + [url]
+
     @classmethod
     def _wait_for_es(cls):
         start_time = time.time()
Index: src/azul/es.py
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/src/azul/es.py b/src/azul/es.py
--- a/src/azul/es.py	(revision 848a45385e8689310693a8e1668074d0fd463770)
+++ b/src/azul/es.py	(date 1696057873429)
@@ -171,7 +171,8 @@
         # error handling, we disable the implicit retries via max_retries=0.
         common_params = dict(hosts=[dict(host=host, port=port)],
                              timeout=timeout,
-                             max_retries=0)
+                             max_retries=0,
+                             connection_class=AzulRequestsHttpConnection)
         if host.endswith('.amazonaws.com'):
             aws_auth = CachedBotoAWSRequestsAuth(aws_host=host,
                                                  aws_region=aws.region_name,
@@ -179,8 +180,6 @@
             return Elasticsearch(http_auth=aws_auth,
                                  use_ssl=True,
                                  verify_certs=True,
-                                 connection_class=AzulRequestsHttpConnection,
                                  **common_params)
         else:
-            return Elasticsearch(connection_class=AzulUrllib3HttpConnection,
-                                 **common_params)
+            return Elasticsearch(**common_params)
Index: test/service/test_manifest_async.py
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/test/service/test_manifest_async.py b/test/service/test_manifest_async.py
--- a/test/service/test_manifest_async.py	(revision 848a45385e8689310693a8e1668074d0fd463770)
+++ b/test/service/test_manifest_async.py	(date 1696057914566)
@@ -181,8 +181,7 @@
             mock_start_execution.reset_mock()
             mock_describe_execution.reset_mock()
 
-        with responses.RequestsMock() as helper:
-            helper.add_passthru(str(self.base_url))
+        with self._requests_mock_helper():
             for fetch in (True, False):
                 with self.subTest(fetch=fetch):
                     execution_id = '6c9dfa3f-e92e-11e8-9764-ada973595c11'
Index: src/azul/logging.py
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/src/azul/logging.py b/src/azul/logging.py
--- a/src/azul/logging.py	(revision 848a45385e8689310693a8e1668074d0fd463770)
+++ b/src/azul/logging.py	(date 1696057794727)
@@ -85,6 +85,9 @@
     actual = [(logger.name, logger.name.startswith(prefix)) for logger in loggers]
     assert actual == expected, actual
     _configure_non_app_logging(get_test_logger(), *loggers)
+    # Responses, the mocking library for Requests, logs a meaningless INFO-level
+    # message ("request.allowed-passthru") for every request it is not mocking.
+    logging.getLogger('responses').setLevel(logging.WARNING)
 
 
 log_format = ' '.join([
```

**Comment by dsotirho-ucsc** [comment]
Spike to profile the slow test with patch in place.

**Comment by hannes-ucsc** [comment]
Don't have time for this now.

**Comment by hannes-ucsc** [comment]
PR #6213 switches ES client to use urllib3. It does not completely eliminate Requests as a dependency because the requests signing still relies on `urllib3.PreparedRequest` to represent the request to be signed. However, and this is a big gain, the `urllib3.Session` session class is not involved in active production code paths anymore. That same PR also eliminates AzulClient's use of Requests in a similar way, but leaning a little more heavily on `urllib3.PreparedRequest`.

I'm still not aware of an easy 3rd-party mocking library for urllib3, similar to what Responses is to Requests.

These changes make this issue more or less obsolete. Moving to Triage so that we can close it as `wontfix` once that PR lands.

**Comment by hannes-ucsc** [comment]
Close `wontfix` when blocker is merged.
